### PR TITLE
hometask_2

### PR DESCRIPTION
**Pull request by izy-13**

### CODE REVIEW

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  static resetIdCounter(): void {
+    this.idCounter = 0;
+  }
+
+  getId(): number {
+    return this.id;
+  }
+
+  use(): void {}
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts]: this method is abstract

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  toString(): string {
+    return `${this.name} - Value: ${this.value.toFixed(2)}, Weight: ${this.weight.toFixed(2)}`
+  }
+
+  compareTo(other: Item): number {
+    if(this.value > other.value) return 1;
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts]: this realization is not optimal, to compare values we can do `return this.value > other.value ? 1 : -1;` to compare string we can use `return this.name.localeCompare(other.name);` 
please. rewrite the method in 3 lines

File: docs/3-building-blocks-of-oop-part-2/hometask/src/ItemWeightComparator.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { ItemComparator } from "./ItemComparator";
+import { Item } from "./Item";
+
+export class ItemWeightComparator implements ItemComparator {
+  compare(first: Item, second: Item): number {
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/ItemWeightComparator.ts:5]: The same comment to realization as in Item

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      return `You use the ${this.name}, dealing ${this.MODIFIER_CHANGE_RATE} points of damage.\nThe ${this.name} breaks.`
+    }
+
+    if(isActive) {
+      this.baseDurability -= this.MODIFIER_CHANGE_RATE;
+      return `You use the ${this.name}, dealing ${this.MODIFIER_CHANGE_RATE} points of damage.`;
+    }
+  }
+
+  polish(): void {}
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts]: do we need this method?

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  toString(): string {
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts:38]: you extended from Item class, which has toString method, can we use it for partially?

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    if(isWillBeBroke) {
+      this.baseDurability -= this.MODIFIER_CHANGE_RATE;
+      return `You use the ${this.name}, dealing ${this.MODIFIER_CHANGE_RATE} points of damage.\nThe ${this.name} breaks.`
+    }
+
+    if(isActive) {
+      this.baseDurability -= this.MODIFIER_CHANGE_RATE;
+      return `You use the ${this.name}, dealing ${this.MODIFIER_CHANGE_RATE} points of damage.`;
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts]: you repeat yourself in this part `You use the ${this.name}, dealing ${this.MODIFIER_CHANGE_RATE} points of damage.`

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    super(name, value, weight);
+    this.baseDamage = baseDamage;
+    this.baseDurability = baseDurability;
+    this.damageModifier = 0;
+    this.durabilityModifier = 0;
+  }
+
+  use(): string {
+    const isWillBeBroke = this.baseDurability - this.MODIFIER_CHANGE_RATE < this.MODIFIER_CHANGE_RATE;
+    const isBroken = this.baseDurability < this.MODIFIER_CHANGE_RATE;
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts]: in task we have `If the effective durability of the Weapon reaches 0 or less, the Weapon will break` but your conditions are not the same,
you take for base this one 
`if (this.baseDurability <= 0) {
      this.baseDurability = 0;
    }
`

then only word with type conversions for `this.baseDurability`
`if (!this.baseDurability) {
return ....}`


`return .....`

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Sword.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import { Weapon } from "./Weapon";
+
+export class Sword extends Weapon {
+  constructor(baseDamage: number, baseDurability: number, value: number, weight: number) {
+    super('sword', baseDamage, baseDurability, value, weight);
+  }
+
+  polish(): void {
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Sword.ts:8]: i see that polish by his logic overrides `getEffectiveDamage` from Weapon, so you should override this method in subclass
take this decission 
`polish(): void {
    this.damageModifier = this.damageModifier + Weapon.MODIFIER_CHANGE_RATE;
  };`
  
  `getEffectiveDamage(): number {
    this.baseDamage = super.getEffectiveDamage();`

    if (this.baseDamage * 100 > 125) {
      this.baseDamage = 1.25
    }
    
    return this.baseDamage;
  }`

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  isSpoiled: boolean = false;
+
+  constructor(name: string, value: number, weight: number, isSpoiled?: boolean) {
+    super(name, value, weight);
+    this.isSpoiled = isSpoiled;
+  }
+
+  use(): string {
+    if(this.isConsumed) return `There's nothing left of the ${this.name} to consume.`;
+    if(this.isSpoiled) return `You consumed the ${this.name}.\nYou feel sick.`;
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts]: you are repeating `You consumed the ${this.name}`

Comment (AlenaBDE) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts]: And also please avoid writing `else if` return in one line with conditions

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { Item } from "./Item";
+
+export class Consumable extends Item {
+  isConsumed: boolean = false;
+  isSpoiled: boolean = false;
+
+  constructor(name: string, value: number, weight: number, isSpoiled?: boolean) {
+    super(name, value, weight);
+    this.isSpoiled = isSpoiled;
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts:9]: in task we have `A newly-created Consumable object should have its isConsumed field set to false.`, i do not see where you change value to `true`

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Pizza.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import { Consumable } from "./Consumable";
+
+export class Pizza extends Consumable {
+  readonly numberOfSlices: number = 0;
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Pizza.ts]: no needs type

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import { Item } from "./Item";
+
+export abstract class Weapon extends Item {
+  readonly MODIFIER_CHANGE_RATE: number = 0.05;
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts]: is MODIFIER_CHANGE_RATE static?

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts (hunk 1 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts (hunk 2 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/ItemWeightComparator.ts (hunk 3 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/ItemWeightComparator.ts:5]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts (hunk 4 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts (hunk 5 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts:38]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Sword.ts (hunk 8 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Sword.ts:8]: Don't cath it all, we need 
image

maybe something like this (see new commit)?

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts (hunk 9 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts (hunk 10 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Consumable.ts:9]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Pizza.ts (hunk 11 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Pizza.ts]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts (hunk 12 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts]: will be updated in next commit

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts [hunk 13]
```diff
@@ -26,16 +26,15 @@ export class Item implements Comparable<Item> {
...
 
   toString(): string {
     return `${this.name} - Value: ${this.value.toFixed(2)}, Weight: ${this.weight.toFixed(2)}`
   }
 
   compareTo(other: Item): number {
-    if(this.value > other.value) return 1;
-    if(this.value < other.value) return -1;
-    if(this.value === other.value && this.name !== other.name) return -1;
+    if(this.value !== other.value) return this.value > other.value ? 1 : -1;
```
Comment (AlenaBDE) [inline docs/3-building-blocks-of-oop-part-2/hometask/src/Item.ts:36]: for future, please do not write in one line some conditions and calculations, this code is not readable

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts (hunk 6 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts]: sorry, I missed it the first time, will be updated

File: docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts (hunk 7 above)
Comment (izy-13) [reply docs/3-building-blocks-of-oop-part-2/hometask/src/Weapon.ts]: sorry, I missed it the first time, will be updated

**Pull request merged**